tests: can: convert CAN tests to the new ztest API

henrikbrixandersen opened the pull request:
Convert the CAN related test suites to use the new ztest API.

This is an alternative to #47496 and #47380 with improved handling of failed prerequisites (based on the work by @yperess in #48768).

These changes were verified on the following boards:
- `lpcxpresso55s16`
- `nucleo_g474re`
- `twr_ke18f`
- `mimxrt1024_evk`
- `frdm_k64f`
- `stm32f3_disco`
- `native_posix`
- `native_posix_64`

Apart from one known issue (see #47014) all tests pass on all boards.

henrikbrixandersen committed:
tests: drivers: can: api: convert to new ztest API

Convert the CAN controller driver API test suite to use the new ztest API.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
This is an alternative to #47496, illustrating the issues encountered when wanting to error-out on test suite prerequisites not being met.

henrikbrixandersen commented on tests/drivers/can/api/src/main.c:
```diff
@@ -1122,42 +1163,49 @@ static void test_filters_preserved_through_bitrate_change(void)
+ */
+ZTEST(can_api, test_prerequisites)
+{
+	zassert_true(device_is_ready(can_dev), "CAN device not ready");
+	zassert_equal(can_dev_init_status, 0, "failed to set loopback mode (err %d)",
+		      can_dev_init_status);
```
If I could just do these asserts in the `can_api_setup()` function and have the entire test suite error-out (not skip, but report an error/failure)...

aaronemassey replied:
So the proposed method here is:

Have a single tiny test suite like `ZTEST_SUITE(can_api_setup)` that only tests the setup success. After that, the `ZTEST_SUITE(can_api)` suite can safely assume the in the setup/before because if the assumption fails and skips the suite, we'll see the failure in the `can_api_setup` suite.

henrikbrixandersen replied:
Which suite would call the setup function? Only one of them can call it and the suite execution order is not guaranteed.

aaronemassey replied:
tldr: 
Yeah, this is more difficult than it should be. What you have right now works,
but it should be easier than this.

Ahh okay. So I decided to pull down the changes and try to make them myself and
I definitely get it now.

In my opinion, this PR is a great case-example for zasserts being supported in
the setup since we can only call `can_set_mode()` once and only once.

My original idea was to have two separate ZTEST_SUITES, one small one with a
single test that verifies the `can_set_mode()` and another that assumes
`can_set_mode()` and then goes along with the other tests. But as mentioned,
`can_set_mode()` can only be called once.

So the `predicate` parameter in the ZTEST_SUITE would be an existing solution to
this problem. Where we run two suites (one suite testing prequisites another
testing other requirements) and selectively choose one suite at a time. But I
don't believe it's intended for such a simple example as this. Let me know if
you'd like me to show you the change here. 

The alternative would be running each one in its own binary, but that seems a
bit much to me as well. I'll raise this and get a lot more attention on this
Monday. Thank you! I really appreciate you finding this example and notifying me
about it.

Would you like me to approve your PR as it stands now?

henrikbrixandersen replied:
> Would you like me to approve your PR as it stands now?

Thanks, but no. I'd rather wait for @yperess' work on being able to use `zassert` in `setup` and convert all the CAN drivers tests in one go.

yperess commented:
> This is an alternative to #47496, illustrating the issues encountered when wanting to error-out on test suite prerequisites not being met.

I'm going to try and crank out a change to make asserts work in setup. Like assume, they'll fail the whole suite and just not run the tests.

henrikbrixandersen committed:
tests: drivers: can: canfd: convert to new ztest API

Convert the CAN-FD controller driver API test suite to use the new ztest
API.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
tests: drivers: can: timing: convert to new ztest API

Convert the CAN controller timing test suite to use the new ztest API.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
tests: drivers: can: utilities: convert to new ztest API

Convert the CAN utilities test suite to use the new ztest API.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
tests: canbus: isotp: implementation: convert to new ztest API

Convert the CAN bus ISO-TP implementation test suite to use the new ztest
API.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
tests: canbus: isotp: conformance: convert to new ztest API

Convert the CAN bus ISO-TP conformance test suite to use the new ztest API.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
I have updated this based on the changes to the ztest framework. Please review.

str4t0m commented on tests/drivers/can/canfd/src/main.c:
```diff
@@ -388,24 +374,25 @@ void test_send_receive_fd(void)
-			 ztest_unit_test(test_send_receive_fd),
-			 ztest_unit_test(test_send_receive_mixed));
-	ztest_run_test_suite(canfd_driver);
+	err = can_set_mode(can_dev, CAN_MODE_LOOPBACK);
```
```suggestion
	err = can_set_mode(can_dev, CAN_MODE_LOOPBACK | CAN_MODE_FD);
```

henrikbrixandersen replied:
Ouch, thanks. Fixed.

str4t0m reviewed (changes requested):
Thanks, with asserts in the setup now we surely won't miss any skipped test cases.
Seems all fine, just need proper canfd setup.
